File: app/src/main/java/utils/Ping.java
package utils;

import java.io.IOException;
import java.net.Inet4Address;
import java.net.Inet6Address;
import java.net.InetAddress;
import java.net.UnknownHostException;

public class Ping {
    public String host;
    public int timeout = 5000;
    public int ttl; // set ttl
    public String ip; // host resolved ip
    public String hoop; // hoop or destination name
    public String hoopIp; // hoop (or destination) ip
    public long ms; // ping delay
    public boolean ttlex; // ttl exceeded
    public int ttlr; // ttl recived (host pinged)
    public int seq; // icmp_seq=387
    public boolean unreachable; // From 10.10.6.1: icmp_seq=1 Destination Port Unreachable
    public boolean reachable; // ping timeout or reachable

    public static InetAddress getHostByName(String host) throws UnknownHostException {
        InetAddress[] aa = InetAddress.getAllByName(host);
        for (InetAddress a : aa) {
            if (a instanceof Inet4Address)
                return a;
            if (a instanceof Inet6Address)
                return a;
        }
        return null;
    }

    public Ping() {
    }

    public Ping(String host) {
        this.host = host;
    }

    public Ping(String host, int ttl) {
        this.host = host;
        this.ttl = ttl;
    }

    public boolean ping() {
        try {
            InetAddress address = PingExt.getHostByName(host);
            if (address == null)
                throw new RuntimeException("Unknown host: " + host);
            ip = address.getHostAddress();
            long now = System.currentTimeMillis();
            if (ttl != 0) {
                boolean r = address.isReachable(null, ttl, timeout);
                ms = System.currentTimeMillis() - now;
                reachable = ms < timeout;
                ttlex = !r;
                hoopIp = host;
            } else {
                reachable = address.isReachable(timeout); // return true even if host is "Destination is Unreachable"
                ms = System.currentTimeMillis() - now;
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        return true;
    }

    public boolean fail() {
        return !reachable;
    }
}

File: app/src/main/java/utils/Ping.kt
package utils

import java.io.IOException

open class Ping {
    @JvmField
    var host: String? = null
    @JvmField
    var timeout = 5000
    var ttl // set ttl
            = 0
    @JvmField
    var ip // host resolved ip
            : String? = null
    @JvmField
    var hoop // hoop or destination name
            : String? = null
    @JvmField
    var hoopIp // hoop (or destination) ip
            : String? = null
    @JvmField
    var ms // ping delay
            : Long = 0
    @JvmField
    var ttlex // ttl exceeded
            = false
    @JvmField
    var ttlr // ttl recived (host pinged)
            = 0
    @JvmField
    var seq // icmp_seq=387
            = 0
    @JvmField
    var unreachable // From 10.10.6.1: icmp_seq=1 Destination Port Unreachable
            = false
    var reachable // ping timeout or reachable
            = false

    constructor()

    constructor(host: String?) {
        this.host = host
    }

    constructor(host: String?, ttl: Int) {
        this.host = host
        this.ttl = ttl
    }

    open fun ping(): Boolean {
        try {
            val address = PingExt.getHostByName(host)
                ?: throw RuntimeException("Unknown host: $host")
            ip = address.hostAddress
            val now = System.currentTimeMillis()
            if (ttl != 0) {
                val r = address.isReachable(null, ttl, timeout)
                ms = System.currentTimeMillis() - now
                reachable = ms < timeout
                ttlex = !r
                hoopIp = host
            } else {
                reachable =
                    address.isReachable(timeout)
                ms = System.currentTimeMillis() - now
            }
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
        return true
    }

    open fun fail(): Boolean {
        return !reachable
    }
}